### ISSUE TITLE
torch.topk is incorrect about the meaning of returned indices

### ISSUE THREAD

**Issue by rodrigodesalvobraz** [pytorch/pytorch#50331]
## 📚 Documentation

[torch.topk documentation](https://pytorch.org/docs/stable/generated/torch.topk.html) says that the function returns a tuple `(values, indices)`, where _indices_ is "the indices of the elements in the original input tensor."

The straightforward interpretation of that sentence is that _indices_ is a collection containing _n_-tuples where _n_ is the number of dimensions in the original tensor _input_. That is, each element _e_ in _indices_ can be used to index the _input_ tensor to obtain the corresponding top-k element.

However, this is incorrect because what _indices_ actually contains is only the index in the _maximized dimension_. They are not the indices in all dimensions, and therefore cannot be used to index the _input_ tensor.

My suggestion for improving the documentation would include the following:

Assuming the original tensor _input_ has shape (m_1, m_2, ..., m_n), `topk(input,dim=i)` returns `(values, indices)` where both _values_ and _indices_ are tensors of shape `(m_1, ..., m_{i-1}, m_{i+1}, ..., m_n)` such that

`values[j_1, ..., j_{i-1}, j_{i+1}, ..., j_n] == max([input[j_1, ..., j_{i-1},  p,    j_{i+1}, ..., j_n] for p in range(input.shape[i]) ])`

and

`indices[j_1, ..., j_{i-1}, j_{i+1}, ..., j_n] == max(range(input.shape[i]), key=lambda p: input[j_1, ..., j_{i-1},  p,    j_{i+1}, ..., j_n])`

using Python's built-in `max` function with the `key` parameter.

**Comment by mruberry** [comment]
Thank you for the suggestion, @rodrigodesalvobraz. I agree the documentation here could be clearer. There are probably other operators dealing with a similar issue. This particular suggestion for an update seems reasonable, but we should work out language that works for other operators with this behavior, too.

**Comment by rodrigodesalvobraz** [comment]
Thank you. Yes, I agree that this aspect is shared with other operators, namely operators that perform reductions such as `torch.sum(input, dim, keepdim=False, *, dtype=None) → Tensor`.

In fact the documentation of `sum` is quite simple:

> Returns the sum of each row of the input tensor in the given dimension dim

so perhaps the documentation of `topk` could be similar:

> A namedtuple of `(values, indices)` is returned, with the values and indices of the largest `k` elements of each row of the `input` tensor in the given dimension `dim`.

**Comment by mruberry** [comment]
Yep, that makes sense.

**Comment by bdhirsh** [comment]
reserving for bootcamp
